File: Bayesian Inference Tools/BUS/BUS_SuS_Python/BUS_SuS_Python/main_example_1_1D.py
"""
---------------------------------------------------------------------------
Example 1: 1D posterior
---------------------------------------------------------------------------
Created by:
Fong-Lin Wu
Felipe Uribe
Luca Sardi

Engineering Risk Analysis Group
Technische Universitat Munchen
www.bgu.tum.de/era
Contact: Antonios Kamariotis (antonis.kamariotis@tum.de)
---------------------------------------------------------------------------
Last Version 2021-03
* Adaptation to new ERANataf class
---------------------------------------------------------------------------
References:
1."Bayesian inference with subset simulation: strategies and improvements"
   Wolfgang Betz et al.
   Computer Methods in Applied Mechanics and Engineering 331 (2018) 72-93.
---------------------------------------------------------------------------
"""

#=================================================================
# import libraries
import numpy as np
import scipy as sp
from scipy.stats import multivariate_normal as mvn
import matplotlib.pylab as plt

# Make sure ERADist, ERANataf classes are in the path
# https://www.bgu.tum.de/era/software/eradist/
from ERADist import ERADist
from ERANataf import ERANataf
from BUS_SuS import BUS_SuS
plt.close('all')

#=================================================================
# define the prior
d           = 1        # number of dimensions (uncertain parameters)
prior_pdf_1 = ERADist('normal','PAR',[0,1])
prior_pdf   = [prior_pdf_1]

# correlation matrix
# R = np.eye(d)   # independent case

# object with distribution information
# prior_pdf = ERANataf(prior_pdf,R)

#=================================================================
# define likelihood
mu_obs    = 5
sigma_obs = 0.2
#
realmin        = np.finfo(np.double).tiny  # to prevent log(0)
likelihood     = lambda theta: mvn.pdf(theta, mu_obs, sigma_obs**2)
log_likelihood = lambda theta: np.log(likelihood(theta) + realmin)

#=================================================================
# find scale constant c
method = 1
# use MLE to find c
if method == 1: 
    f_start = np.log(mu_obs + realmin)
    fun     = lambda lnF: -log_likelihood(np.exp(lnF))
    MLE_ln  = sp.optimize.fmin(func=fun, x0=f_start)
    MLE     = np.exp(MLE_ln)          # likelihood(MLE) = 1
    c       = 1/likelihood(MLE)
        
# some likelihood evaluations to find c
elif method == 2:
    K  = 5e3                          # number of samples
    x  = np.random.normal(size=(d,K)) # samples in standard space
    # likelihood function evaluation
    L_eval = np.zeros(K,1)
    for i in range(K):
        L_eval[i] = likelihood(np.transpose(x[:,i]))
    c = 1/max(L_eval)                 # Ref. 1 Eq. 34
    
# use approximation to find c
elif method == 3:
    print('This method requires a large number of measurements')
    m = len(mu_obs)
    p = 0.05
    c = 1/np.exp(-0.5*sp.stats.chi2.ppf(p,df=m))     # Ref. 1 Eq. 38
        
else:
    raise RuntimeError('Finding the scale constant c requires -method- 1, 2 or 3')

#=================================================================
# BUS-SuS step
N  = int(3e3)    # number of samples per level
p0 = 0.1         # probability of each subset

print('\nBUS with SUBSET SIMULATION: \n')
h, samplesU, samplesX, logcE, sigma = BUS_SuS(N, p0, c, log_likelihood, prior_pdf)

#=================================================================
# organize samples and show results
nsub     = len(h.flatten())  # number of levels + final posterior
u1p, u0p = list(),list()
x1p, pp  = list(),list()
#
for i in range(nsub):
   # samples in standard
   u1p.append(samplesU['total'][i][:,0])  
   u0p.append(samplesU['total'][i][:,1])
   # samples in physical
   x1p.append(samplesX[i][:,0])
   pp.append(samplesX[i][:,1])

#=================================================================
# reference and BUS solutions
mu_exact    = 4.81     
sigma_exact = 0.196
cE_exact    = 2.36e-6

# show results
print('\nExact model evidence = ', cE_exact)
print('Model evidence BUS-SuS = ', np.exp(logcE))
print('\nExact posterior mean = ', mu_exact)
print('Mean value of samples = ', np.mean(x1p[-1]))
print('\nExact posterior std = ', sigma_exact)
print('Std of samples = ', np.std(x1p[-1]),'\n')

File: Bayesian Inference Tools/1. BUS/BUS_SuS_Matlab/main_example_1_1D.m
%% example 1: 1D posterior
%{
---------------------------------------------------------------------------
By:                Date:          Topic:
Fong-Lin Wu        July.2019      
---------------------------------------------------------------------------
Current version 2021-03
* use of log-evidence logcE
---------------------------------------------------------------------------
References:
1."Bayesian inference with subset simulation: strategies and improvements"
   Wolfgang Betz et al.
   Computer Methods in Applied Mechanics and Engineering 331 (2018) 72-93.
---------------------------------------------------------------------------
%}
clear; clc; close all; 

% add path to ERADist and ERANataf classes
% Source: https://www.bgu.tum.de/era/software/eradist/
% addpath('../../../ERA_Dist/ERADistNataf_MATLAB/')

%% initial data
d = 1;      % number of dimensions (number of uncertain parameters)

%% prior PDF for theta
prior_pdf = ERADist('normal','PAR',[0,1]);

%% likelihood PDF (mixture of gaussians in 2D)
s = RandStream.create('mt19937ar','seed',12001);   % fixed the seed to get the same data points
RandStream.setGlobalStream(s);
%
mu             = 5;
sigma          = 0.2;
likelihood     = @(theta) normpdf(theta,mu,sigma);
log_likelihood = @(theta) log(likelihood(theta)+realmin);   % realmin to avoid Inf values in log(0)

%% find scale constant c
method = 1;
switch method
   % use MLE to find c
   case 1  
      f_start = log(mu);
      fun     = @(lnF) -log(likelihood(exp(lnF))+realmin);
      options = optimset('MaxIter',1e7,'MaxFunEvals',1e7);
      MLE_ln  = fminsearch(fun,f_start,options);
      MLE     = exp(MLE_ln);   % likelihood(MLE) = 1
      c       = 1/likelihood(MLE);
      
   % some likelihood evaluations to find c
   case 2
      K  = 5e3;                  % number of samples      
      u  = randn(d,K);           % samples in standard space
      x  = u;       % samples in physical space
      
      % likelihood function evaluation
      L_eval = zeros(K,1);
      for i = 1:K
         L_eval(i) = likelihood(x(i));
      end
      c = 1/max(L_eval);    % Ref. 1 Eq. 34
      
   % use approximation to find c
   case 3
      fprintf('This method requires a large number of measurements\n');
      m = length(mu);
      p = 0.05;
      c = 1/exp(-0.5*chi2inv(p,m));    % Ref. 1 Eq. 38
      
   otherwise
      error('Finding the scale constant c requires -method- 1, 2 or 3');
end

%% BUS-SuS step
N  = 3e3;    % number of samples per level
p0 = 0.1;    % probability of each subset

fprintf('\nBUS with SUBSET SIMULATION: \n');
[h, samplesU, samplesX, logcE, sigma] = BUS_SuS(N, p0, c, log_likelihood, prior_pdf);

%% extract the samples
nsub = length(h);
u1p  = cell(nsub,1);  u0p = cell(nsub,1);
x1p  = cell(nsub,1);  pp  = cell(nsub,1);
for i = 1:nsub
   % samples in standard
   u1p{i} = samplesU.total{i}(:,1);             
   u0p{i} = samplesU.total{i}(:,2); 
   % samples in physical
   x1p{i} = samplesX{i}(:,1);
   pp{i}  = samplesX{i}(:,2);
end

%% reference and BUS solutions
mu_exact    = 4.81;      % for x_1
sigma_exact = 0.196;     % for x_1
cE_exact    = 2.36e-6;

% show results
fprintf('\nExact model evidence = %g', cE_exact);
fprintf('\nModel evidence BUS-SuS = %g\n', exp(logcE));
fprintf('\nExact posterior mean = %g', mu_exact);
fprintf('\nMean value of samples = %g\n', mean(x1p{end}));
fprintf('\nExact posterior std = %g', sigma_exact);
fprintf('\nStd of samples = %g\n\n', std(x1p{end}));
%%END